docs/arch/Proposals/adr-2504-bulkupdate.md

```diff
@@ -0,0 +1,155 @@
+- **Trying to add an existing element** – Trying to add an existing element to a resource would fail
+- **Trying to replace an element that don't exist** – Trying to replace an element that don't exist would fail
+- **Update call resulting into an invalid resource** – Should error if the PATCH update forms an invalid resource. Is there a case where one resource would result in an invalid resource and other resources of the same type won't?
+- **Partial updates** – Resources with valid requests should get updated. The failed resources should be logged with the reason.
+- **Logging** – Should have the audit log for each bulk-update call with the resource IDs of the items that have been updated successfully. In case of error, should provide the valid reason with failed to update resource IDs.
+
+
+###  High‑Level Flow
+
+```mermaid
+sequenceDiagram
+    participant C as Client
+    participant G as API Gateway
+    participant Q as Job Queue
+    participant W as Job Workers
+    participant DB as Data Store
+    participant AU as Audit Svc
+    C->>G: PATCH /Patient/$bulk-update?…\nPrefer: respond-async
+    G-->>C: 202 Accepted\nContent-Location: /jobs/123
+    G->>Q: enqueue job 123
+    loop until complete
+        W->>Q: dequeue chunk
+        W->>DB: apply FHIR Patch (batch)
+        W->>AU: write AuditEvent
+    end
+    W->>G: job 123 complete
+    C->>G: GET /jobs/123
+    G-->>C: 200 OK\nItems-Updated: N
+```
+
+###  In-depth work Flow
+* [Bulk Update workflow](https://github.com/microsoft/fhir-server/blob/main/docs/flow%20diagrams/bulk-update.md)
+
+### Job pattern
+
+| Step | Behaviour |
+|------|-----------|
+| Submit | Server returns **202 Accepted** with `Content-Location: {pollingUrl}` |
+| Processing | Job queued; workers update resources in chunks (≥ 1 000 items/sec target) |
+| Complete | Polling URL returns **200 OK** when done. Headers: `Items-Updated`, `X-Error-Count` |
+| Outputs | `OperationOutcome` (no PHI) |
+| Cancel | `DELETE {pollingUrl}` ⇒ **202 Accepted** → job status = `cancelled` |
+| Audit & Metrics | Each update emits an `AuditEvent` (`update\|bulk-update`) and increments metric |
+| Events | Updated items will cause an event in Azure FHIR Events |
+
+---
+
+### Status
+Proposed
+
+---
+
+## Consequences
+
+| Benefit | Consequence |
+|---------|-------------|
+| Mass partial updates without full re‑imports | Higher write load on datastore |
+| Detailed per‑resource audit trail | Increased `AuditEvent` volume |
+
+---
+
+## Estimated performance numbers based on previous runs on Update operation
+2000 resources per seconds (For soft $bulk-delete, we roughly achieved 1600 reosurces per seconds)
```

Comment:
2000 max, it is probably unreasonable to assume we will hit that target.

Reply:
yes, is it safe to say 1600? (based on bulk-delete numbers)

Reply:
I'm hesitant to list performance numbers in a planning doc, but 1600 seems doable.